### PR TITLE
Add Rails test mappings (minitest & rspec)

### PR DESCRIPTION
**Pull request by erikw**
This is a collaborative PR that will add mappings for rails projects.

For rails (and really Ruby in general), the two most common testing frameworks without doubt are minitest and rspec. 

References on recommended test file naming convention:
* Minitest: `_test` suffix https://github.com/rubocop/minitest-style-guide#file-naming
* RSpec: `_spec` suffix https://github.com/rubocop/rspec-style-guide#view-spec-file-name

Example repos of the file structure:
* Minitest: https://github.com/awesome/minitest-rails-example
* RSpec: https://github.com/eliotsykes/rspec-rails-examples

As a start this PR includes some fixtures based on a large rails project that I work on

## TODOs
- [x] @erikw Add fixtures for minitest
- [x] @erikw Add fixtures for rspec
  * I just copied minitest and ran inside the copied dir: `$ find . -type f -exec rename 's/_test.rb/_spec.rb/' {} \;`
- [x] Create mappings
  * Ideally it should be posible to jump from source -> test (choose when different options e.g. functional or integration of same class), and to go from any test -> source
  * The rules that I currently use for this plugin can be seen in #15. While these rules works for my minitest project, they are very verbose and should be made general. Probably there should not be individual rules for `app/model`, `app/controller` going to `test/unit/model`, `app/controller`, but ideally something more general like `app/*` -> `test/unit/*` and so on. 
  * As the only difference between the minitest and rspec naming is the suffix, the rules should be made general to match on `.*_(test|spec).rb` instead of duplicating rules per framework.
- [x] @erikw Test the final mappings in a large real-world rails project. This is something that I can do!

This PR is made collaborative, so please push to this branch just @rgroli !

### CODE REVIEW

**Comment by rgroli** [comment]
@erikw Great work, thank you! I'll have a look in a few days!

**Comment by rgroli** [comment]
@erikw I set up the mappings as good as I could and tried to keep the already existing rails mappings.
Beside the mappings between various "app" and "test" I added mappings for all kinds of files in "app" for various contexts.

So it would be cool if you could test it in your real-life project and don't hesitate to tell me if I messed up 😄 

I also enhanced the window for picking the right file and added a shortcut column to directly pick via single key.

What's missing now is automated tests, which I'll be working on next.

**Comment by erikw** [comment]
@rgroli 🚀 nice!

I will make Tine to test this properly next week:)

File: lua/other-nvim/builtin/mappings.lua [hunk 1]
```diff
@@ -5,61 +5,127 @@ M.laravel = {
...
+		pattern = "/app/(.*)/(.*).rb",
+		target = {
+			{ context = "test", target = "/test/unit/%1/%2_test.rb" },
+			{ context = "test", target = "/test/functional/%1/%2_test.rb" },
+			{ context = "test", target = "/test/functional/%2_test.rb" },
+			{ context = "test", target = "/test/integration/%1/%2_test.rb" },
+			{ context = "test", target = "/test/integration/%2_test.rb" },
+
+			{ context = "test", target = "/test/%1/%2_test.rb" },
+			{ context = "test", target = "/test/%1/%2_test.rb" },
```
Comment (erikw) [inline lua/other-nvim/builtin/mappings.lua:45]: Duplicated line from above?

File: lua/other-nvim/builtin/mappings.lua [hunk 2]
```diff
@@ -5,61 +5,127 @@ M.laravel = {
...
+			{ context = "test", target = "/test/%2_test.rb" },
+
+			{ context = "test", target = "/spec/unit/%1/%2_spec.rb" },
+			{ context = "test", target = "/spec/functional/%1/%2_spec.rb" },
+			{ context = "test", target = "/spec/functional/%2_spec.rb" },
+			{ context = "test", target = "/spec/integration/%1/%2_spec.rb" },
+			{ context = "test", target = "/spec/integration/%2_spec.rb" },
+
+			{ context = "test", target = "/spec/%1/%2_spec.rb" },
+			{ context = "test", target = "/spec/%1/%2_spec.rb" },
```
Comment (erikw) [inline lua/other-nvim/builtin/mappings.lua:57]: Duplicated line from just above?

**Comment by erikw** [comment]
@rgroli this is really shaping up! The window picker looks great as well.

In general the most important bindings works. There are just a few tweaks more. Using here `lua/spec/fixtures/rails-minitest` as the example (some files were missing in the fixtures, now added):
* `db/migrations/20220923181300_add_email_to_user.rb` does not find the test
   * `test/migrations/20220923181300_add_email_to_user_test.rb`
* `app/models/feature.rb` does not find the two corresponding controllers and single channel:
  * `app/controllers/api/v1/feature_controller.rb`
  * `app/controllers/api/v2feature_controller.rb`
  * `app/channels/api/v1/feature_channel.rb`
* `app/controllers/api/v1/feature_controller.rb` does find its unit test but not integration or functional:
   * `test/functional/api/v1/feature_controller_test.rb`
   * `test/integration/api/v1/feature_controller_test.rb`
* `lib/user_helper.rb` does not find its unit test (same for `lib/core_ext/string.rb`)
   * `test/unit/lib/user_helper_test.rb` 

All other mappings seemed to work great though! Thanks for your improvements :)

**Comment by rgroli** [comment]
Alright cool, I try to finish the missing mappings and automatic testing this week. Hopefully I can merge this week. 
Another test from you before that would be greatly appreciated 😄

**Comment by erikw** [comment]
Of course, I will happily test more iterations. Just ping me!

**Comment by erikw** [comment]
Many thanks @rgroli for the update!

I re-run the tests that did not work from last time:

@rgroli this is really shaping up! The window picker looks great as well.

In general the most important bindings works. There are just a few tweaks more. Using here `lua/spec/fixtures/rails-minitest` as the example (some files were missing in the fixtures, now added):
* `db/migrations/20220923181300_add_email_to_user.rb` does not find the test
   * `test/migrations/20220923181300_add_email_to_user_test.rb`
      * ✅ fixed
* `app/models/feature.rb` does not find the two corresponding controllers and single channel:
  * `app/controllers/api/v1/feature_controller.rb`
      * ✅ fixed
  * `app/controllers/api/v2/feature_controller.rb`
      * ✅ fixed
  * `app/channels/api/v1/feature_channel.rb`
     * ❌ not found, but also maybe not the worst, most important is that the channel finds it's own test (which it does!)
* `app/controllers/api/v1/feature_controller.rb` does find its unit test but not integration or functional:
   * `test/functional/api/v1/feature_controller_test.rb`
      * ✅ fixed
   * `test/integration/api/v1/feature_controller_test.rb`
      * ✅ fixed
* `lib/user_helper.rb` does not find its unit test (same for `lib/core_ext/string.rb`)
   * `test/unit/lib/user_helper_test.rb` 
      * ✅ fixed

I think this looks really good now, I will be happy to use this a as-is for the rails production projects that I work on!



## Note
Would it be possible to add configuration to make the selection window larger, so that one can see a bit more of the path? I can create a separate issue for this!

Screen Shot 2022-10-18 at 13 32 34

**Comment by rgroli** [comment]
thanks for the feedback! I'm on vacation right now and will continue next week..

**Comment by rgroli** [comment]
@erikw I fixed the last mapping issue and finally merged 😄  

For #17 I will create another branch to work on.

Thanks again for working with me on the issue - very much apprechiated! 🚀 🚀

**Comment by erikw** [comment]
Yay 🚀 🚀 

Thanks for fixing that last mapping too, I confirmed that it works.

It was a pleasure working with you, and now it will be a pleasure to work with rails in nvim <3

**Pull request merged**